src/mono/mono/eventpipe/ep-rt-mono.c

```diff
@@ -2857,6 +2857,433 @@ ep_rt_mono_write_event_ee_startup_start (void)
+			// if it were the only type batched in the whole event.  Bail
+			return -1;
+		}
+	}
+	if (p_type_logger->m_n_bulk_type_value_byte_count + cb_val > k_max_bytes_type_values)
+	{
+		// Although this type fits into the array, its size is so big that the entire
+		// array can't be logged via ETW. So flush the array, and start over by
+		// calling ourselves--this refetches the type info and puts it at the
+		// beginning of the array.  Since we know this type is small enough to be
+		// batched into an event on its own, this recursive call will not try to
+		// call itself again.
+		ep_rt_mono_fire_bulk_type_event(p_type_logger);
+		return ep_rt_mono_log_single_type(p_type_logger, mono_type, type_id);
+	}
+	mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "p_val added at index: %d\ntype_id: %d\nmodule_id: %d\ntype_name_id: %d\ntype: %d\nc_type_parameters: %d", p_type_logger->m_n_bulk_type_value_count, p_val->fixed_sized_data.type_id, p_val->fixed_sized_data.module_id, p_val->fixed_sized_data.type_name_id, mono_underlying_type, p_val->c_type_parameters);
+	for (int i = 0; i < p_val->c_type_parameters; i++)
+	{
+		mono_trace (G_LOG_LEVEL_DEBUG, MONO_TRACE_DIAGNOSTICS, "type: %d\n", ((MonoType*)p_val->rg_type_parameters[i])->type);
+	}
+
+	// The type fits into the batch, so update our state
+	p_type_logger->m_n_bulk_type_value_count++;
+	p_type_logger->m_n_bulk_type_value_byte_count += cb_val;
+	return p_type_logger->m_n_bulk_type_value_count - 1;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// High-level method to batch a type and (recursively) its type parameters, flushing to
+// ETW as needed.  This is called by ep_rt_mono_log_type_and_parameters_if_necessary.
+//
+// Arguments:
+//      * mono_type - MonoType to be logged
+//      * type_id - Unique type identifier of mono_type to batch
+
+void
+ep_rt_mono_log_type_and_parameters (BulkTypeEventLogger *p_type_logger, MonoType *mono_type, intptr_t type_id)
+{
+	// Batch up this type.  This grabs useful info about the type, including any
+	// type parameters it may have, and sticks it in m_rg_bulk_type_values
+	uint32_t i_bulk_type_event_data = ep_rt_mono_log_single_type (p_type_logger, mono_type, type_id);
+	if (i_bulk_type_event_data == -1)
+	{
+		// There was a failure trying to log the type, so don't bother with its type
+		// parameters
+		return;
+	}
+
+	// Look at the type info we just batched, so we can get the type parameters
+	BulkTypeValue *p_val = &p_type_logger->m_rg_bulk_type_values[i_bulk_type_event_data];
+
+	// We're about to recursively call ourselves for the type parameters, so make a
+	// local copy of their type handles first (else, as we log them we could flush
+	// and clear out m_rg_bulk_type_values, thus trashing p_val)
+	uint32_t c_params = p_val->c_type_parameters;
+	intptr_t rg_type_parameters[c_params];
+	MonoType *rg_mono_type_parameters[c_params];
```

Comment:
I'm surprised this didn't flag a compiler error (maybe it did; I didn't look at the build logs). You'll want to just use the `const` max number of param types since this is stack allocated.

Reply:
variable length arrays work in C99 (and on compilers we support).  But yea, there's no real reason to use them here.

Reply:
Changed to fixed length array with size `k_size_of_type_parameter_array` (32)